### ISSUE TITLE
Do Puppet web apps need 

### ISSUE THREAD

**Issue by warpech** [Starcounter/Starcounter#3562]
Currently, all the sample apps (and [implicit standalone mode](https://github.com/Starcounter/Starcounter/blob/3e6d5b2cec36e12e09c06d1968b8f8709c33d37d/src/Starcounter/Middleware/PartialToStandaloneHtmlProvider.cs)) have the following code that forces Polymer to use shimmed Shadow DOM instead of Shady DOM.

```html
<script>
        window.Polymer = window.Polymer || {};
        window.Polymer.dom = "shadow";
 </script>
```

@tomalec, @miyconst - do we really need this? If I understand correctly, there is no actual benefit for us.

**Comment by miyconst** [comment]
This is only needed for performance. Native Shadow DOM works much faster in Chrome, compare to the shims.

Note that the puppet apps work in FF and IE, which no where near support native Shadow DOM.

**Comment by warpech** [comment]
> Native Shadow DOM works much faster in Chrome, compare to the shims.

Polymer does not use Shadow DOM shims. It uses Shady DOM. Is native Shadow DOM noticeably faster than Shady DOM?

**Comment by miyconst** [comment]
At least it was told so by Polymer guys on the Polymer Summit. And it was one of their recommendations, to always turn on Shadow DOM if available.

**Comment by tomalec** [comment]
>  if available.

The question is: Is the pollyfilled Shadow DOM (if native is not available) faster than Shady DOM?

I suspect it's not.

It would be cool to have at least one benchmark test in Launcher suite. So we could compare those two.

On the other hand supporting shady may also require us to maintain non-encapsulated CSS

**Comment by warpech** [comment]
> It would be cool to have at least one benchmark test in Launcher suite. So we could compare those two.

I made some performance tests using a fork of [PolymerLabs/js-repaint-perfs](https://github.com/PolymerLabs/js-repaint-perfs). Below are results:

**Warning. This can kill your browser**

- ["naive" Polymer with webcomponents-lite.min.js](http://warpech.github.io/js-repaint-perfs/polymer/index.html)
- ["naive" Polymer with webcomponents.min.js](http://warpech.github.io/js-repaint-perfs/polymer/index-shadow.html)
- ["naive" Polymer with webcomponents.min.js and forced Shadow DOM](http://warpech.github.io/js-repaint-perfs/polymer/index-shadow-forced.html)
- ["optimised" Polymer with webcomponents-lite.min.js](http://warpech.github.io/js-repaint-perfs/polymer/opt.html)
- ["optimised" Polymer with webcomponents.min.js](http://warpech.github.io/js-repaint-perfs/polymer/opt-shadow.html)
- ["optimised" Polymer with webcomponents.min.js and forced Shadow DOM](http://warpech.github.io/js-repaint-perfs/polymer/opt-shadow-forced.html)
- ["complex" Polymer with webcomponents-lite.min.js](http://warpech.github.io/js-repaint-perfs/polymer/complex.html)
- ["complex" Polymer with webcomponents.min.js](http://warpech.github.io/js-repaint-perfs/polymer/complex-shadow.html)
- ["complex" Polymer with webcomponents.min.js and forced Shadow DOM](http://warpech.github.io/js-repaint-perfs/polymer/complex-shadow-forced.html)

I didn't make the naive/optimised/complex performance suites; I took them from PolymerLabs and just added two Shadow DOM related variants to them.

- | Chrome | Firefox | Safari | Edge
--- | --- | --- | --- | ---
complex `wc-lite.min.js` | 60 fps / 9.54M | **71 fps** | **94 fps** | **305 fps**
complex `wc.min.js` | 60 fps / 9.54M | 40 fps | 64 fps | 294 fps
complex `wc.min.js` forced Shadow DOM | **62 fps / 9.54M** | 43 fps | 65 fps | 294 fps

The numbers above were tested with Polymer 1.4. Edge was tested with dev tools open.

I cannot explain why there is no difference when forcing Shadow DOM. I was expecting to see at least some difference. 

Anyway, I think this supports my claim from OP, unless someone has a better explanation?

**Comment by miyconst** [comment]
@warpech thank you for the nice numbers, I am also surprised. I agree, that now we can disable it.

**Comment by warpech** [comment]
Removed forced Shadow DOM from `PartialToStandaloneHtmlProvider`, which was automatically applied to KitchenSink.

I think Launcher should go next, and after that standalone modes in all other apps.

**Comment by miyconst** [comment]
>I think Launcher should go next, and after that standalone modes in all other apps.

Most of our apps use implicit standalone mode, means we don't have to change the Polymer settings in all of the apps. Those apps which don't use implicit standalone mode should be changed to use it, because now there is no way to distinguish which mode to use.

**Comment by warpech** [comment]
> Most of our apps use implicit standalone mode, means we don't have to change the Polymer settings in all of the apps. Those apps which don't use implicit standalone mode should be changed to use it, because now there is no way to distinguish which mode to use.

Yes, totally this! Even Launcher could use implicit standalone mode.

This is the list of black sheep apps: https://github.com/search?utf8=%E2%9C%93&q=import+title+puppet+starcounter+user%3AStarcounter+user%3AStarcounterSamples+user%3AStarcounterPrefabs+user%3AJuicy+user%3AStarcounter-Jack+user%3APuppetJs&type=Code&ref=searchresults

**Comment by warpech** [comment]
@tomalec, @miyconst I plan to do this change today. Any suggestions how to check that nothing broke down?

**Comment by miyconst** [comment]
The only automated way we have is the KitchenSink. Others - check `starcounter-debug-aid` and layout editor.

**Comment by tomalec** [comment]
This seem to break existing ::shadow CSS selectors in 
https://github.com/StarcounterPrefabs/Launcher/blob/2886d9b746917eb953a051601d40178f5077085e/src/Launcher/wwwroot/Launcher/css/style.css

**Comment by warpech** [comment]
Thanks for noticing. Reported as a separate issue: https://github.com/StarcounterPrefabs/Launcher/issues/235
